### PR TITLE
Parse errormsgs in retryable status codes

### PR DESCRIPTION
**Pull request by pree-dew**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/5536

### CODE REVIEW

**Comment by pree-dew** [comment]
@MrAlias PR #5537 had incorrect email address, that's why CLA was failing in order to fix it I have to change git history, which is not recommended in contribution guidelines so created this new PR and closing the old one.

File: exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go:180]: We shouldn't lose the status code here. `%w` would allow unwrapping the error, and retrieving it back while retaining additional information.
The `%w` change is also happening in https://github.com/open-telemetry/opentelemetry-go/pull/5535

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:27]: Why the import order change?

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:23]:
```suggestion
- Pass the underlying error rather than a generic retry-able failure in `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`. (#5541)
```

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:199]:
```suggestion
```

This comment makes sense within the PR, but outside its context, it doesn't that much.

**Review by dmathieu — COMMENTED** [review]
If we do this for metrics, we should also do it for traces and logs.

File: exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go:180]: This function is getting used while writing a response, `fmt.Fprintln(w, e.Error())` and as per earlier code it is sending status code in response while in actual collector response we only get error message not status code.

I completely agree on using %w when we are returning `error` but here we are returning a string.

Please correct if I am missing something here.

Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go:180]: Understood, I will make the change and update the other test cases accordingly, as there might be a few test cases which will break if status code with start showing in response.

Thanks, it makes sense.

Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetricgrpc/internal/otest/collector.go]: Addressed.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: Addressed.

File: CHANGELOG.md
Comment (pree-dew) [reply CHANGELOG.md]: Addressed.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: Addressed. Let me know if it needs any improvement

**Comment by pree-dew** [comment]
@dmathieu Is there anything missing/incorrect in this PR that needs a fix? Happy to accomodate that.

**Review by dmathieu — COMMENTED** [review]
We need to do this for traces too.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go:293]: I agree, if the parsing has to happen in this struct then we should change the name from `errMsg` to `body`

**Comment by pree-dew** [comment]
@dmathieu Added the error message handling in traces and logs also.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: This is addressed by using a status package used on collector also to construct response body in case of error

File: CHANGELOG.md
Comment (pree-dew) [reply CHANGELOG.md]: Addressed

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:45]:
```suggestion
- Pass the underlying error rather than a generic retry-able failure in `go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp`, `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp` and `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp`. (#5541)
```

File: exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go:195]: Should this test be added to log and trace too?

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:310]:
```suggestion
	rErr.errMsg = body
	if st, ok := status.FromError(errors.New(body)); ok {
		rErr.errMsg = fmt.Sprintf("rpc error: code = %s desc = %s", st.Code(), st.Message())
	}
```

Body is already allocated, while sprintf allocates a new string in memory. No need to do that allocation if the content isn't ok.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:294]:
```suggestion
// throttle delay contained in headers.
```

This comment is about the behavior of `retryableError`, that's where it should be.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (dmathieu) [inline exporters/otlp/otlpmetric/otlpmetrichttp/client.go:285]:
```suggestion
// retryableError represents a request failure that can be retried.
//
// If the `errMsg` attribute is not empty, it will be used as the error message.
```

File: exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go:195]: Sorry, it's a miss from my side. Traces already had a test case for the use case there I have changed how the error message will reflect now. Added for logs.

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: Makes complete sense, addressed

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go]: addressed

File: exporters/otlp/otlpmetric/otlpmetrichttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlpmetric/otlpmetrichttp/client.go:285]: addressed

**Review by dmathieu — APPROVED** [review]

**Comment by dmathieu** [comment]
There is a conflict that should be easily resolvable. And a merge requires 2 approvals.

**Comment by pree-dew** [comment]
@pellared Can you please help in reviewing/approving this? We have customers using otel via http and this PR will help in debugging pipeline issues as it solves https://github.com/open-telemetry/opentelemetry-go/issues/5536 .
Sorry for this but it will be a great help.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client.go:286]: This should be the wrapped error, not the error message. Details about the underlying error and ability to interact with the error using the Go tooling are lost when this is not stored as the wrapped error.

The [`Unwrap`](https://pkg.go.dev/github.com/pkg/errors#Unwrap) and [`As`](https://pkg.go.dev/github.com/pkg/errors#As) semantics should be added to the `retryableError` to support use cases where this needs to be unwrapped.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client.go:302]: This will never be satisfied:

https://github.com/grpc/grpc-go/blob/v1.65.0/status/status.go#L96-L127

This function needs to accept an `error` that is going to be wrapped for this to succeed in determining the status from the error.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: @MrAlias Thanks for this suggestion, I think it makes sense. I have accommodated these changes, let me know if it looks okay.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client.go:188]:
```suggestion
```

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client.go:207]:
```suggestion
			if respStr := strings.TrimSpace(respData.String()); respStr != "" {
				// Include response for context.
				e := errors.New(respStr)
				rErr = newResponseError(resp.Header, e)
			} else {
				rErr = newResponseError(resp.Header, nil)
			}
```

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: @MrAlias That's actually very good point, we don't have to pre-render it. 

FIxed.

Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: @MrAlias All comments that you have given are really good. I have happily accommodated those and pushed the fix.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client_test.go:719]: This is testing `err` is not nil. It is using `exporterErr` as a message: https://pkg.go.dev/github.com/stretchr/testify/assert#Error

```suggestion
		assert.ErrorIs(t, err, exporterErr)
```

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client_test.go:722]: https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorAs

```suggestion
		assert.ErrorAs(t, err, &retryErr)
```

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client_test.go:727]: This seems like it is testing the wrapping semantics of Go itself by checking multiple unwraps here. Is that correct?

If so, I think the assertion above `assert.ErrorIs` should be sufficient. It calls `Unwrap` recursively until the base `exportErr` is returned and should be sufficient to validate the `Unwrap` implementation.

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client.go:287]: Outdated now.

```suggestion
// throttle delay contained in headers. The returned error wraps wrapped if it is not nil. 
```

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client_test.go]: Basically, wanted to test the `As` implementation and then verify the error. Changed it to `ErrorIs`.

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client_test.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client_test.go]: Changed it to ErrorContains, as ErrorIs will not work here, sample link for explanation 
https://go.dev/play/p/5wzdTmyy1pD

File: exporters/otlp/otlplog/otlploghttp/client.go
Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client.go]: Addressed.

File: exporters/otlp/otlplog/otlploghttp/client_test.go
Comment (MrAlias) [inline exporters/otlp/otlplog/otlploghttp/client_test.go:724]: This is a no-op test. `retryErr` was asserted to be `err` on the previous line.

I would use `assert.ErrorIs(t, err, exporterErr)` in place of `ErrorContains` on line 719. This will effectively test the same thing `ErrorContains` tests (the error string is guaranteed to be equal if the identity is asserted) while also testing the unwrap functionality.

Comment (pree-dew) [reply exporters/otlp/otlplog/otlploghttp/client_test.go:724]: I understand that `ErrorIs` can be used but here it will fail because `err` and `exporterErr` are different instances, unlike with `retryErr` where `ErrorAs` has passed reference of wrapped err and `ErrorIs` is true in next assert. 

`exporterErr` is created to be passed in exporter result but it is not directly wrapped internally in the flow like  **CASE 3** in this link https://go.dev/play/p/ap14MIXRk21

Here it is following **CASE 2** in the above link, that's why `ErrorIs` will fail here.

Let me know if there is a gap in this.

Comment (MrAlias) [reply exporters/otlp/otlplog/otlploghttp/client_test.go:724]: Right, makes sense :+1:

**Comment by MrAlias** [comment]
@dmathieu PTAL.

**Review by dmathieu — APPROVED** [review]

**Comment by dmathieu** [comment]
CI is failing after this merge (for some reason, it didn't fail before, so that may be a flake).
https://github.com/open-telemetry/opentelemetry-go/actions/runs/9987613497/job/27602574063

**Comment by pree-dew** [comment]
@MrAlias @dmathieu Thank you for helping and guiding throughout the issue, appreciate it.

**Pull request merged**
